### PR TITLE
[blog] tested article RAW

### PR DESCRIPTION
**Pull request by TomasVotruba**
It is time, ref: https://twitter.com/hrachcz/status/816940948114571264

### CODE REVIEW

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+layout: post
+title: "Why articles with code examples should be CI tested"
+perex: "I know many great articles, that go right to the point and when I use code examples, they work. But when I recommend these article to people I mentor, I realize they are 2 years old and probably not working any more. From hero to zero. Today I will show how to keep them alive lot longer with minimal effort."
+author: 1
+lang: en
+---
+
+Do you know [Awesome lists](https://github.com/sindresorhus/awesome/)? If not, got check them. They collect best sources about some topic. When I try to learn something new, I start on Github looking for "awesome <technology>". I recommend at least briefly checking them.
+ 
+The idea behind Awesome list is to have source that:
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: "source" => "sources"? There is "are" in the list below...

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
+---
+layout: post
+title: "Why articles with code examples should be CI tested"
+perex: "I know many great articles, that go right to the point and when I use code examples, they work. But when I recommend these article to people I mentor, I realize they are 2 years old and probably not working any more. From hero to zero. Today I will show how to keep them alive lot longer with minimal effort."
+author: 1
+lang: en
+---
+
+Do you know [Awesome lists](https://github.com/sindresorhus/awesome/)? If not, got check them. They collect best sources about some topic. When I try to learn something new, I start on Github looking for "awesome <technology>". I recommend at least briefly checking them.
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: "about some topic" => "about certain topics"?

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+One day I decided to learn more about Doctrine. Documentation looked like a manual for experts rather than something I could learn from. I was also curious **how people use Doctrine in real applications, how to overcome performance issues, some cool features and pro tips**.
+ 
+I was already familiar with [Awesome PHP](https://github.com/ziadoz/awesome-php) by [ziadoz](https://github.com/ziadoz), so I looked for "Awesome Doctrine". 
+
+0 results. Really? Why nobody made this? It's so obvious this would be useful.
+
+Ah, it's my job then. And the joyful hell started.
+
+
+### Many Source on Many Versions
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: sources?

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+ 
+I was lucky to found many articles about Doctrine. One about Filters, Events or Criteria. But when I tried to use the code, it often didn't work. After digging I found out there was version 1.0, which was completely different.
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: lucky to find

One about Filters, Events or Criteria. => One about Filters, others about Events or Criteria.

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+0 results. Really? Why nobody made this? It's so obvious this would be useful.
+
+Ah, it's my job then. And the joyful hell started.
+
+
+### Many Source on Many Versions
+ 
+I was lucky to found many articles about Doctrine. One about Filters, Events or Criteria. But when I tried to use the code, it often didn't work. After digging I found out there was version 1.0, which was completely different. 
+
+> Tip: if you write post about software, at least mention its version &ndash; even if it has only single version.
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: Tip: When you write a post about software, mention the version you're referring to &ndash; even if it only has one at the moment.

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+### What to put in the list?
+
+I've decided to focus on sources released in that year. When articles and Doctrine version are the same - Doctrine 2.4 - it will be great source to learn from.
+
+Idea was good, [List](https://github.com/TomasVotruba/awesome-doctrine) was done. I was happy until...
+
+### ...Doctrine 2.5 was out!
+
+So now every from 20 sources in the list got bit deprecated. Thank you!
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: So now each of the 20 sources on the list got a bit deprecated.

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+Oh, so that's why nobody made it in the first place :).
+
+Now I also understand why programmers hate new versions of software and want to stick with version they already know. It makes sense in such conditions.
+
+
+## "Awesome Symfony"  
+
+Before I realized It makes no sense to make list of sources, because next year I could drop most of them, I make [Awesome list for Symfony](https://github.com/Pehapkari/awesome-symfony-education).
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: "It" => "it"

to make a list

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+Idea was good, [List](https://github.com/TomasVotruba/awesome-doctrine) was done. I was happy until...
+
+### ...Doctrine 2.5 was out!
+
+So now every from 20 sources in the list got bit deprecated. Thank you!
+
+Oh, so that's why nobody made it in the first place :).
+
+Now I also understand why programmers hate new versions of software and want to stick with version they already know. It makes sense in such conditions.
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: Not completely true... I for one like new versions even though they require some work. But I get the point. Maybe "why many/most programmers"?

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+If I get back to the **useful source** idea from the beginning.
+
+- **are up-to-date with modern technology**
+- **are the best in the field**
+- **are easy to learn by beginners**
+
+
+To make this happen, I would have to create "Awesome * List" every year.
+To make that happen, each article would have to be rewritten with each new version.
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: each article would have to be checked for compatibility with each new version and updated if needed.

(you don't need to completely rewrite it... most of the time...)

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+That would mean around 50 articles on Doctrine every year. **And all this work just to keep status quo**. In big communities like Symfony and Laravel, this happens, but I still consider too much wasted work (constructive ideas coming bellow).
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: I still consider it

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+So every year sources are more outdated, useful source then it used to be in the day of release of article. Writing article in such environment would be as useless as writing 100% test coverage for Christmas campaign microsite.
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: the first sentense doesn't make much sense to me... maybe sth like this?

So sources are useful upon their release but become more and more outdated every year.

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+
+
+To make this happen, I would have to create "Awesome * List" every year.
+To make that happen, each article would have to be rewritten with each new version.
+
+That would mean around 50 articles on Doctrine every year. **And all this work just to keep status quo**. In big communities like Symfony and Laravel, this happens, but I still consider too much wasted work (constructive ideas coming bellow).
+
+So every year sources are more outdated, useful source then it used to be in the day of release of article. Writing article in such environment would be as useless as writing 100% test coverage for Christmas campaign microsite. 
+
+Thus, motivation to write article software is getting low, even when software is being released. - I call this *Know How Sharing Lag*.
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: to write software articles

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+That would mean around 50 articles on Doctrine every year. **And all this work just to keep status quo**. In big communities like Symfony and Laravel, this happens, but I still consider too much wasted work (constructive ideas coming bellow).
+
+So every year sources are more outdated, useful source then it used to be in the day of release of article. Writing article in such environment would be as useless as writing 100% test coverage for Christmas campaign microsite. 
+
+Thus, motivation to write article software is getting low, even when software is being released. - I call this *Know How Sharing Lag*.  
+
+
+## This Sounds like Legacy Code 
+
+Let's say we have application with legacy code. It brings me money and I want keep it alive and growing as much longer as possible.
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: I want to keep it... as long as possible

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+Thus, motivation to write article software is getting low, even when software is being released. - I call this *Know How Sharing Lag*.  
+
+
+## This Sounds like Legacy Code 
+
+Let's say we have application with legacy code. It brings me money and I want keep it alive and growing as much longer as possible. 
+ 
+...
+
+Mm, I would write tests and start refactoring?
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: would => should?

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+
+Could this be possible to integrate to a blog or website?
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: to integrate into a blog

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+
+### Dream Big
+
+It would have to be:
+
+- **integrated in blog**, because another external source would deprecate - thanks Sculpin
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: Somehow the first time I read this I thought you were ironic - thanking Sculpin for being deprecated. :-)

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+- **composer supported** - thanks Github Pages and Travis  
+- open-source hosted, so the author won't burn out on yearly fixes - thanks Github Pages
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: no bold here?

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+- **CI supported** - thanks Travis
+- **tested daily** - thanks Travis Cron Jobs (Beta)
+
+
+This idea was created in late 2015 with no solution ahead. I want to thank [Jáchym Toušek](https://twitter.com/enumag) for consulting this idea and making [the first prototype](https://github.com/enumag/enumag.cz/commit/3efc82717b9965bb19a2609e4caddc0c5467552d).  
+
+
+And that why and how *tested articles* were born.
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: that's

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+
+
+
+### On this Blog by Default!
+
+As you can see, there are few tested articles already:
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: a few?

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+
+<img src="/assets/images/posts/2017/tested-article/covered.png">
+
+This is how tested article looks like in [single commit](https://github.com/pehapkari/pehapkari.cz/commit/85b69950b32c39b9e972582720a23a18a1adc4be). It will last for years, will work on Symfony 4 and maybe further.
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: This is what a tested article looks like

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+
+### On this Blog by Default!
+
+As you can see, there are few tested articles already:
+
+<img src="/assets/images/posts/2017/tested-article/covered.png">
+
+This is how tested article looks like in [single commit](https://github.com/pehapkari/pehapkari.cz/commit/85b69950b32c39b9e972582720a23a18a1adc4be). It will last for years, will work on Symfony 4 and maybe further. 
+
+**Feel free to send one.** We'll make sure it will make it into 2018 :)
```
Comment (enumag) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: Missing dot

maybe "into 2018 and beyond"

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md (hunk 8 above)
Comment (TomasVotruba) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: I see, my original statement was too strict. Thank you

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md (hunk 9 above)
Comment (TomasVotruba) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: I was talking rather about major versions, but you are right.

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md (hunk 11 above)
Comment (TomasVotruba) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: I got lost in my imagination bit there

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md (hunk 16 above)
Comment (TomasVotruba) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: Oh, old memory. Rather Statie here :D

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md (hunk 19 above)
Comment (TomasVotruba) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: Good one! I didn't know this

File: source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md (hunk 21 above)
Comment (TomasVotruba) [inline source/_posts/2017/2017-01-20-why-articles-with-code-examples-should-be-CI-tested.md]: "beyond" is very nice, thanks

**Pull request merged**